src/liballoc/task.rs

```diff
@@ -60,9 +60,11 @@ impl<W: Wake + Send + Sync + 'static> From<Arc<W>> for RawWaker {
 fn raw_waker<W: Wake + Send + Sync + 'static>(waker: Arc<W>) -> RawWaker {
     // Increment the reference count of the arc to clone it.
     unsafe fn clone_waker<W: Wake + Send + Sync + 'static>(waker: *const ()) -> RawWaker {
-        let waker: Arc<W> = Arc::from_raw(waker as *const W);
-        mem::forget(Arc::clone(&waker));
-        raw_waker(waker)
+        Arc::incr_strong_count(waker as *const W);
+        RawWaker::new(
+            waker as *const (),
+            &RawWakerVTable::new(clone_waker::<W>, wake::<W>, wake_by_ref::<W>, drop_waker::<W>),
+        )
```

Comment:
Hm, so this specific use site would actually benefit from returning an Arc from incr_strong_count, or perhaps a better name would then be `Arc::clone_from_raw` which would return a Arc, but not by conceptually taking ownership of the Arc in the raw pointer.

That API also seems easier to understand.

Reply:
It does not use/return an `Arc`?

Reply:
It previously called raw_waker (which takes that Arc), avoiding the duplicated call to RawWaker::new

Reply:
A simple edit here could instead be to do:

```rust
raw_waker(Arc::from_raw(waker as *const W))
```
This would be equivalent to doing a clone, which is the way the code was prior to this patch. I think with the methods proposed in this patch we have nice a set of composable operations for using arc for custom refcounts:
- `Arc::into_raw` cast an arc into a pointer
- `Arc::from_raw` cast a pointer back into an arc
- `Arc::incr_strong_count` increment the refcount through a pointer
- `Arc::decr_strong_count` decrement the refcount through a pointer

Reply:
No, that's not equivalent to cloning, that's consuming the Arc conceptually owned by the raw pointer. Unless you meant clone_from_raw, but that doesn't jive with the rest of your comment.